content/cloud-hosting/azure/azure-service-bus.md

```diff
@@ -0,0 +1,254 @@
+
+เมื่อสัปดาห์ที่แล้ว ได้มีโอกาศรับงานต่อจากเพื่อนมา เพื่อจะนำระบบ Message Queue มาใช้งานจริงใน project ที่กำลังดำเนินการอยู่ ซึ่งเพื่อนก็ทำการบ้านมาเป็นอย่างดี มีตัวอย่างการใช้งาน Azure Service Bus มาให้
+
+แต่อย่างไรก็ตาม สุดท้ายผมติดปัญหาการนำมาใช้จริง เนื่องจากการ implement ด้วย Azure Service Bus สามารถทำงานได้บน Cloud แต่ถ้าเราจะทดสอบระบบ ระหว่างที่ debug ในเครื่องตัวเอง, การจะเปิด Queue บน Cloud สำหรับทุกคน นั่นค่อนข้างลำบากและมีค่าใช้จ่าย จึงจำเป็นต้องหาหนทาง ในการทำให้ทำงานใน local ได้, และ deploy ไปใช้งานร่วมกับ Cloud Service ได้ โดยใช้ code ชุดเดียวกัน โดยในที่สุด ผมก็ทำได้ โดยเก็บ source code ไว้ที่นี่ นะคับ
+
+ ```
+https://github.com/nakornttss/azureservicebusrabmitmqamqp
+ ```
+
+![Azure Service Bus and RabbitMQ demo source code](https://miro.medium.com/max/1400/1*NCg_LziF6RT5i9-2qoy-WA.png)
+
+จากการศึกษา พบ Mesage Queue ที่น่าสนใจคือ RabbitMQ ซึ่งใช้ protocal ชื่อ AMQP (https://www.amqp.org/) เป็นโปรโตคอล พื้นฐานสำหรับจัดการคิว ซึ่งเป็นตัวเดียวกันกับที่ Azure Service Bus ใช้
+
+แต่อย่างไรก็ตาม แต่ละโปรแกรม จะมี library ในการจัดการคิวของตนเอง ดังนี้
+
+- Azure Service Bus ศึกษาได้จาก https://docs.microsoft.com/en-us/dotnet/api/overview/azure/service-bus
+- RabbitMQ จะให้ใช้ dotnet add package RabbitMQ.Client โดยดูตัวอย่างได้จาก https://www.rabbitmq.com/tutorials/tutorial-one-dotnet.html
+
+จากข้อมูลข้างต้น ทำให้เราไม่สามารถใช้ code ชุดเดียวกัน ในการรันทั้ง local และ cloud ได้ จึงจำเป็นจะต้องใช้ library อื่นมาเป็นสะพานกลาง เพื่อให้ code เดียวกัน ใช้งานกับทั้งคู่ได้ ซึ่งผมศึกษาเพิ่ม พบว่า มีตัวที่น่าสนใจคือ http://azure.github.io/amqpnetlite/
+
+จากนั้น ผมก็ทดสอบ เอามาใช้งาน ปรากฏว่า ติดปัญหาเพิ่มอีกนิดหน่อย คือว่า Azure Service Bus ใช้ protocal version 1.0 แต่ RabbitMQ ใช้ protocal version 0.9.x ซึ่ง พอไม่ตรงกัน เลยต้องมีการ configure ให้ RabbitMQ สามารถใช้ 1.0 ได้ด้วย ซึ่งผมได้จัดการ สร้าง dockerfile ขึ้นมา โดย inherit ต่อมาจาก official 3.9 และได้จัดแจงเปิด service เพิ่มเติม เพื่อให้ RabbitMQ ของผม สามารถดู dashboard ใน local ได้ด้วย จึงได้ผล ออกมาเป็น dockerfile และ docker compose ดังนี้
+
+dockerfile
+
+ ```dockerfile
+FROM rabbitmq:3.9
+RUN rabbitmq-plugins enable --offline rabbitmq_management rabbitmq_amqp1_0
+RUN set eux; \
+# make sure the metrics collector is re-enabled (disabled in the base image for Prometheus-style metrics by default)
+rm -f /etc/rabbitmq/conf.d/management_agent.disable_metrics_collector.conf; \
+# grab "rabbitmqadmin" from inside the "rabbitmq_management-X.Y.Z" plugin folder
+# see https://github.com/docker-library/rabbitmq/issues/207
+cp /plugins/rabbitmq_management-*/priv/www/cli/rabbitmqadmin /usr/local/bin/rabbitmqadmin; \
+[ -s /usr/local/bin/rabbitmqadmin ]; \
+chmod +x /usr/local/bin/rabbitmqadmin; \
+apt-get update; \
+apt-get install -y --no-install-recommends python3; \
+rm -rf /var/lib/apt/lists/*; \
+rabbitmqadmin --version
+EXPOSE 15672
+EXPOSE 5672
+EXPOSE 5671
+  ```
+
+docker-compose.yaml
+
+ ```yaml
+version: "3.2"
```

Comment:
I think Docker Compose file must have an indent, could you please make the two spaces correctly.
Please check this [example](https://github.com/dotnetthailand/dotnetthailand.github.io/blob/main/content/storage/sql-server/docker-compose-for-sql-server.md?plain=1#L93-L137).

Reply:
Fixed